src/parser.rs

```diff
@@ -24,18 +24,16 @@ pub enum ChunkParseError {
 /// This enables sorting all chunks by id for reassembly
 /// Chunk format: 003OF008andthenbase64payloadhere
 pub fn parse(chunk: &str) -> Result<EncodedChunk, ChunkParseError> {
-    let chunk_id = match chunk[..3].parse::<u16>() {
-        Ok(i) => i,
-        Err(_) => return Err(ChunkParseError::IdMissing),
-    };
-    if chunk[3..5] != String::from("OF") {
+    let chunk_id = chunk[..3]
+        .parse::<u16>()
+        .map_err(|_| ChunkParseError::IdMissing)?;
```

Comment:
I love the `map_err`, and want to adopt it. But  This trips me up: `map_err` seems to make chunk_id into a `Result<u16, ChunkParseError`, which means for my overall parser function to work (returning a Result<EncodedChunk, ChunkParseError>), I have to keep on going with the "confusion" of having chunk_id potentially not an integer yet.
Note the added value here of the match is the `return` bit, so that it's not just `chunk_id` is potentially a result obj, but the function early exited. This operator doesn't do that, because it is an expression, not a statement?

So yeah, if I understand correctly, to make this function work with the PR, I need to add at the end before the `Ok()` a kind of `if let e = Err(chunk_id) { return e;}` and then the same for `chunk_total` etc? Or is there a more elegant solution?

Reply:
Whoops seems I missed the little `?` which is the _actual_ `try` operator, and confused with `map_err`. All good, I guess!